Network event update message

rickychien opened the pull request:
Fixes #196, support network event update message will enable status information properly and tests has added.

Default status is hidden and will be visible once client receives network event update message.

<img width="1232" alt="screen shot 2016-09-16 at 11 48 14 am" src="https://cloud.githubusercontent.com/assets/958165/18574881/df475ff8-7c03-11e6-9139-37df17f79116.png">

- [x] http version
- [x] status
- [x] status text
- [x] total time

rickychien commented:
@linclark please take a look thanks!

linclark commented on devtools/client/webconsole/new-console-output/reducers/messages.js:
```diff
@@ -42,6 +42,17 @@ function messages(state = new MessageState(), action) {
+      let updateMessage = action.message;
+
+      return state.withMutations(function (record) {
+        record.set("messagesById", messagesById.map((message, key) => {
```
If there are lots of messages, then this could be a performance issue. I can see two potential options here.

1. Use the actor as the ID rather than generating a numeric one.
2. Have a separate index which maps actor to ID.

@bgrins do you know if there is any reason not to use the actor as the ID? Can actor strings potentially be reused during the same session?

bgrins commented:
We could also add a stub for a network event update by listening for [networkEventUpdate](https://dxr.mozilla.org/mozilla-central/source/devtools/shared/webconsole/client.js#170) in the stub generator test: https://github.com/devtools-html/gecko-dev/blob/console-frontend/devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/browser_webconsole_update_stubs_network_event.js#L29

bgrins commented on devtools/client/webconsole/new-console-output/actions/messages.js:
```diff
@@ -39,6 +40,20 @@ function messageAdd(packet, idGenerator = null) {
 
+function messageUpdate(packet, idGenerator = null) {
+  return (dispatch) => {
+    if (idGenerator == null) {
```
Chatted with Lin about this.  Here's what I'm thinking:

1) Don't worry about passing in a custom idGenerator here, just call prepareMessage directly
2) Modify the id generation in prepareMessage to pass the packet into getNextID https://github.com/devtools-html/gecko-dev/blob/console-frontend/devtools/client/webconsole/new-console-output/utils/messages.js#L32
3) In getNextID, use `packet && packet.actor` if it's available, otherwise use `"" + this.messageId++` as before (https://github.com/devtools-html/gecko-dev/blob/console-frontend/devtools/client/webconsole/new-console-output/utils/id-generator.js#L16)
4) Delete getCurrentId in the id generator - it doesn't appear to be used and also it would be returning potentially inconsistent things after this change

With these changes, after calling `let message = prepareMessage(packet)` here, the message will have the same ID as the one that needs to be updated.

Then later on in the reducer when 'swapping' the old and new messages, you could compare on the "id" field instead of actor.

rickychien replied:
Did you mean that (3) will be used an actor string instead of messageId when ```packet && packet.actor``` is available? So it would exist two different type of id (actor and numeric id) in the meantime.

And I don't get it how do we remove idGenerator parameter in prepareMessage. 
https://github.com/devtools-html/gecko-dev/blob/console-frontend/devtools/client/webconsole/new-console-output/utils/messages.js#L32
We still use an instance of idGenerator here and as you mentioned to pass a packet like  ```idGenerator.getNextId(packet)``` so that we must call ```let message = prepareMessage(packet, idGenerator)``` as usual.

Is there anything I misunderstood?

bgrins replied:
I meant delete only the getCurrentId function on the idGenerator object (not the id genertor itself, or calls to getNextId): https://github.com/devtools-html/gecko-dev/blob/baeb3a5a0463c1222b515ffb660a57ed5efcba64/devtools/client/webconsole/new-console-output/utils/id-generator.js#L19-L21.  AFAICT that getCurrentId function is never used.

bgrins replied:
> @bgrins Did you mean that (3) will be used an actor string instead of messageId when packet && packet.actor is available? So it would exist two different type of id (actor and numeric id) in the meantime. It seems like a workaround with two inconsistent id types and I just want to confirm that is exactly what you want.

Yes, exactly.  They will both be strings though.  I know that they are using different mechanism to generate them, but I think that's OK since we want to always use the same id for packets that are referencing the same actor so auto-incrementing doesn't work for them.

There's an alternate way to do the same thing where we could add logic in prepareMessage to not call the idGenerator at all if the 'actor' property is available - that'd be fine with me too

rickychien commented:
@bgrins Yeah, we can do it just to generate a stub for networkUpdateEvent but it is useless for testing NetworkEventMessage component.

For testing NetworkEventMessage component, it doesn't need to know what event behind the UI because eventually we'll receive a networkInfo which already updated itself with new coming  networkEventUpdate packet. NetworkEventMessage is merely aware of message props changing rather than a networkUpdateEvent.

So I think it's fine to hardcode a fake response within networkEvent stub (that's what I did in my PR [1]) to test NetworkEventMessage component.

[1] https://github.com/devtools-html/gecko-dev/pull/283/commits/2e63623f10e4045935df0e9ba5fc827c5ae41261#diff-05dd566ef061ae5938e5e71f6d1e5892R54

linclark commented:
@rickychien If possible, I'd rather not hardcode any stubs. Instead, we should be generating the stubs by making the same requests the system would be. This ensures that if anything changes in the backend, our stub is updated automatically.

Is there something that makes it hard to generate the stub this way?

rickychien commented:
@linclark After rebasing to latest commit, I noticed that there is a ```npm test``` error as follow

```
var mockStore = configureStore([thunk]);
                ^

TypeError: configureStore is not a function
    at Object.<anonymous> (mozilla-central/devtools/client/webconsole/new-console-output/test/actions/messages.test.js:12:19)
```

Do you have any idea?

linclark commented:
Did you run npm install?

rickychien commented:
It works correctly after removing ```.default``` from ```const configureStore = require("redux-mock-store").default;```

rickychien commented:
@linclark yes, I ran ```npm install``` and it failed after ```npm test```.

linclark commented:
That's interesting because I had to add that in to make it stop failing. I'm going to follow up on IRC to see what versions you have of the dependencies and whether they are different from mine. That should help to target the issue.

linclark commented on devtools/client/webconsole/webconsole.js:
```diff
@@ -3404,7 +3411,11 @@ WebConsoleConnectionProxy.prototype = {
+      if (this.webConsoleFrame.NEW_CONSOLE_OUTPUT_ENABLED) {
+        this.dispatchMessageUpdate(networkInfo);
+      } else {
+        this.webConsoleFrame.handleNetworkEventUpdate(networkInfo, packet);
```
There is some logic in `handleNetworkEventUpdate` that we'll want to maintain. For example, the logic below is helpful for mochitest tests:

```
    // For unit tests we pass the HTTP activity object to the test callback,
    // once requests complete.
    if (this.owner.lastFinishedRequestCallback &&
        networkInfo.updates.indexOf("responseContent") > -1 &&
        networkInfo.updates.indexOf("eventTimings") > -1) {
      this.owner.lastFinishedRequestCallback(networkInfo, this);
    }
```

rickychien replied:
done!

linclark replied:
Sorry, I should have been more clear.

We want to have that code run in both the old and the new frontend. It is currently running in the old frontend because we're calling `handleNetworkEventUpdate`. The place that you had put it before meant it would run once in new frontend and twice in old frontend. We want it to only run once in both frontends. So if you put that code inside the if statement for the new frontend, it should work.

rickychien commented:
@linclark I've updated my tests to get rid of hardcoding stubs. I just generate the same fake network message to overwrite previous networkInfo so that last message stub will be ultimate state.
For generating network event stub is a little bit of tricky and maybe there is another better solution, but now it works to me.

linclark commented on devtools/client/webconsole/new-console-output/test/fixtures/stubs/networkEvent.js:
```diff
@@ -14,169 +14,883 @@ let stubPackets = new Map();
+  "type": "log"
+}));
+
+stubPreparedMessages.set("GET request", new NetworkEventMessage({
```
The keys here should be different for each object. That way you can access each object individually. This probably means you need to figure out a way to distinguish between the different packets coming in and give them meaningful keys.

rickychien replied:
I agree with that. There is a bit tricky here to give each object an individual key and I notice that ```responseContent``` always present at the end of networkEventUpdate. So I assume the order of these packets are predictable.

rickychien commented on devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/head.js:
```diff
@@ -22,20 +22,22 @@ const { prepareMessage } = require("devtools/client/webconsole/new-console-outpu
   let prepared = prepareMessage(message, {getNextId: () => "1"});
   return `
-stubPreparedMessages.set("${key}", new ConsoleMessage(${JSON.stringify(prepared, null, "\t")}));
+stubPreparedMessages.set("${key}", new ConsoleMessage(${JSON.stringify(prepared, null, 2)}));
```
ok! I though it's just a tiny coding style changes which should not be a significant impact. I'm ok with this and I'd like to follow the same style and keep using "\t"

rickychien commented on devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/head.js:
```diff
@@ -59,9 +61,59 @@ function formatNetworkStub(key, packet) {
+`;
+}
+
+function formatNetworkEventUpdateStub(key, packet) {
```
This logic copied from [1] along with a bit of changes. I think we're unable to reuse it directly since that ```_onNetworkEventUpdate()``` doesn't return a networkInfo object.
I did the same thing in previous patch for network event message which copied logic from the same place to generate a fake networkInfo [2].

[1] https://github.com/devtools-html/gecko-dev/blob/console-frontend/devtools/shared/webconsole/client.js#L129-L168
[2] https://github.com/devtools-html/gecko-dev/blob/console-frontend/devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/head.js#L38-L61

linclark replied:
> Besides, I did the same thing in previous patch for network event message which copied logic from the same place to generate a fake networkInfo [2].

Right, I didn't notice that was what had been done until after the review. We should remove that too.

I had to dig around a bit. I didn't realize how complicated the handling of the network stuff was on the client.

I think that if you use `toolbox.target.activeConsole._client.addListener` here (instead of `toolbox.target.client.addListener`), then you shouldn't need the code. That will listen for the event emitted by the web console client, rather than the actor. I think that should solve the problem.

rickychien replied:
Cool! That is pretty useful 👍 

I realized your above example doesn't work since it accesses to debuggerClient. Fortunately, I figured out ```toolbox.target.activeConsole.on()``` can listen to correct event from webConsoleClient itself and solve this issue.

It's a bit difficult for me to find out how to access a particular instance in toolbox such as webConsoleClient. I'll keep in mind to use this trick to access toolbox instance.

rickychien commented on devtools/client/webconsole/new-console-output/test/fixtures/stub-generators/browser_webconsole_update_stubs_network_event.js:
```diff
@@ -24,12 +24,23 @@ add_task(function* () {
+      toolbox.target.client.addListener(`${TARGET}Update`, (type, res) => {
+        stubs.packets.push(formatPacket(keys[0], res));
+        stubs.preparedMessages.push(formatNetworkEventUpdateStub(keys[0], res));
+        if(++i === 8) {
```
I'll add a comment and make it as a const instead.

rickychien commented:
@linclark issues are addressed and please take a look again thanks!

rickychien commented on devtools/client/webconsole/new-console-output/utils/messages.js:
```diff
@@ -29,7 +29,7 @@ function prepareMessage(packet, idGenerator) {
     packet = packet.set("repeatId", getRepeatId(packet));
   }
-  return packet.set("id", idGenerator.getNextId());
+  return packet.set("id", packet.actor || idGenerator.getNextId());
```
done

rickychien commented on devtools/client/webconsole/new-console-output/actions/messages.js:
```diff
@@ -13,6 +13,7 @@ const { IdGenerator } = require("devtools/client/webconsole/new-console-output/u
 
 const {
   MESSAGE_ADD,
+  MESSAGE_UPDATE,
```
I still think it's kind of a general message update and can be applied to all messages who wants to update itself through the messageId.

rickychien replied:
After discussing on IRC, I'll rename it to NETWORK_MESSAGE_UPDATE

rickychien commented:
Indeed, the update events are more challenging than previous one and the code needs to be improved to match current architecture. I'm glad to see network event works as expected without problems!
@linclark Patch has been updated, please review it again.

linclark commented:
Sorry for the delay in responding. For the next couple of days we're going to be focusing on turning the new console output on in Nightly, but I will come back to this after that.

rickychien commented on devtools/client/themes/webconsole.css:
```diff
@@ -154,7 +154,7 @@ a {
 
 .message-flex-body > .message-body {
-  display: block;
+  display: flex;
```
I'm not sure why we split original ```span.message-body ``` into 3 levels and set ```display: block``` caused network layout broken. Thus, I put it back and it  looks great for me now.

rickychien commented:
@linclark It seems there are so many patches landed recently which introduced lots of conflicts after rebasing.

Here is what I fixed:
* I noticed there is a new message component introduced and my patch has been updated again to match the new architecture.
* Fixed layout broken after message component introduced.
* There is an intermittent problem that sometimes network status doesn't show up. After investigating and comparing these networkEvent and networkEventUpdate packets, it's a bug causes networkEventUpdate received before networkEvent. And I noticed there is a new batchActions to delay networkEvent. So it should be used by networkEventUpdate too.
* Fixed ```npm test``` failures.

linclark commented:
@rickychien unfortunately it looks like the `stubPackets` are still all using the same key. I'm not sure how that error was reintroduced. I think I'd have to give it another full review before committing it.

rickychien commented:
@linclark Nice catch! 

I wasn't aware of that last time and I just differentiated the key of ```stubPreparedMessages``` but forgot to do that in ```stubPackets```.

Issues are addressed and fixed stub id for ```formatNetworkEventStub``` and I've checked stub/networkEvent.js again through my eyes.
